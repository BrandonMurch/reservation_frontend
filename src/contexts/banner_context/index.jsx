import React, {
  createContext, useContext, useReducer,
} from 'react';
import PropTypes from 'prop-types';
import Banner, { bannerTypes } from 'general_components/banner';

const BannerContext = createContext();

const BannerContextProvider = function BoilerPlateForBannerContextProvider({
  children,
  value,
}) {
  const initialObject = {
    type: null,
    message: null,
  };

  const reducer = ((state, action) => {
    if (action.clear) {
      return initialObject;
    }
    return { type: action.type, message: action.message };
  });

  const [{ type, message }, dispatchBanner] = useReducer(reducer, initialObject);

  const setBanner = (newType, newMessage) => {
    dispatchBanner({ type: newType, message: newMessage });
    setTimeout(() => {
      dispatchBanner({ type: null, message: null });
    }, 10000);
  };

  return (
    <BannerContext.Provider value={value || setBanner}>
      {message !== null ? <Banner type={type} message={message} /> : null}
      {children}
    </BannerContext.Provider>
  );
};

BannerContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
  value: PropTypes.func,
};

BannerContextProvider.defaultProps = {
  value: null,
};

const useBannerContext = function CustomHookToConsumeBannerContext() {
  return useContext(BannerContext);
};

export {
  BannerContext, useBannerContext, BannerContextProvider, bannerTypes,
};
