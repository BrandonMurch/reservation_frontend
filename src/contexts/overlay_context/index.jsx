import React, {
  createContext, useContext, useState,
} from 'react';
import PropTypes from 'prop-types';

const OverlayContext = createContext();

const OverlayContextProvider = function BoilerPlateForOverlayContextProvider({
  children,
}) {
  const [overlay, setOverlay] = useState(null);

  return (
    <OverlayContext.Provider value={setOverlay}>
      {overlay}
      {children}
    </OverlayContext.Provider>
  );
};

OverlayContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

const useOverlayContext = function CustomHookToConsumeOverlayContext() {
  return useContext(OverlayContext);
};

export {
  useOverlayContext, OverlayContextProvider,
};
