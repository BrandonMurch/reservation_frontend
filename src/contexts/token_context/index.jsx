import React, { createContext, useContext, useRef } from 'react';
import PropTypes from 'prop-types';

const TokenContext = createContext();

const TokenContextProvider = function BoilerPlateForTokenContextProvider({
  defaultValue,
  children,
}) {
  const token = useRef(defaultValue);

  const tokenObject = {
    setToken: (newToken) => {
      token.current = newToken;
    },
    get getToken() {
      return token.current;
    },
  };

  return <TokenContext.Provider value={tokenObject}>{children}</TokenContext.Provider>;
};

TokenContextProvider.propTypes = {
  defaultValue: PropTypes.string,
  children: PropTypes.node.isRequired,
};

TokenContextProvider.defaultProps = {
  defaultValue: null,
};

const useTokenContext = function CustomHookToConsumeTokenContext() {
  return useContext(TokenContext);
};

export { TokenContext, useTokenContext, TokenContextProvider };
