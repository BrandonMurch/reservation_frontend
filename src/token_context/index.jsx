import { createContext } from 'react';

const tokenContext = createContext({
  jwt: undefined,
  setJwt: (jwt) => {
    this.jwt = jwt;
  },
});

export default tokenContext;
