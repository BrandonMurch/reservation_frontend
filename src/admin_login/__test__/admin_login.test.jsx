// Dependencies
import React from 'react';
import { fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { unmountComponentAtNode } from 'react-dom';
import { create } from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { mockFetch, renderWithRouter } from 'test_utils';
import { TokenContextProvider } from '../../contexts/token_context';

// Components
import AdminLogin from '../index';

const fillInInput = async function enterTextIntoInput(element) {
  await fireEvent.focus(element);
  await fireEvent.change(element, { target: { value: 'String' } });
  await fireEvent.blur(element);
};

const fillOutForm = async function fillOutAdminLoginForm(component) {
  const username = component.getByLabelText(/Username/i);
  const password = component.getByLabelText(/Password/i);
  const submit = component.getByRole('button', { name: 'Submit' });
  await act(async () => {
    await fillInInput(username);
    await fillInInput(password);
    await fireEvent.click(submit);
  });
};

describe('<AdminLogin />', () => {
  let component;
  let mockSetErrorFunction;
  let mockSetMessageFunction;
  let fetchSpy;
  let container = null;

  const renderAdminLogin = function boilerplateForRenderingAdminLogin() {
    return renderWithRouter(
      <TokenContextProvider>
        <AdminLogin setError={mockSetErrorFunction} setMessage={mockSetMessageFunction} />
      </TokenContextProvider>,
      { route: '/admin-login' },
    );
  };

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    mockSetErrorFunction = jest.fn();
    mockSetMessageFunction = jest.fn();
  });

  afterEach(() => {
    if (fetchSpy) {
      fetchSpy.mockClear();
    }
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('should match snapshot', () => {
    const tree = create(
      <TokenContextProvider>
        <Router>
          <AdminLogin setError={mockSetErrorFunction} setMessage={mockSetMessageFunction} />
        </Router>
      </TokenContextProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render email and password inputs', () => {
    component = renderAdminLogin();

    const emailInput = component.getByLabelText(/Username/i);
    expect(emailInput).toBeInTheDocument();
    const passwordInput = component.getByLabelText(/Password/i);
    expect(passwordInput).toBeInTheDocument();
  });

  it('should render a submit button', () => {
    component = renderAdminLogin();

    const submitButton = component.getByRole('button', { name: 'Submit' });
    expect(submitButton).toBeInTheDocument();
  });

  it('should successfully redirect on login', async () => {
    const body = {
      token: 'this is a token',
    };

    fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => mockFetch(200, body));

    component = renderAdminLogin();

    await fillOutForm(component);

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(component.history.location.pathname).toEqual('/admin');
  });

  it('should set error when server returns an error', async () => {
    fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => mockFetch(400));
    component = renderAdminLogin();

    await fillOutForm(component);

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const errorMessage = component.getByText(/Something went wrong/i);
    expect(errorMessage).toBeInTheDocument();
  });
});
