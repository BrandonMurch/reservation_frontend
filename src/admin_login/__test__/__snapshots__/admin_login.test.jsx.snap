// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminLogin /> should match snapshot 1`] = `
<div
  className="background"
>
  
  
  <div
    className="container"
  >
    <form
      className="container"
      data-testid="form"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="inputGroup"
      >
        <label
          className="hiddenLabelText"
          htmlFor="username"
        >
          Username:
        </label>
        <input
          className="input"
          id="username"
          name="username"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          pattern={null}
          placeholder="Username"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="inputGroup"
      >
        <label
          className="hiddenLabelText"
          htmlFor="password"
        >
          Password:
        </label>
        <input
          className="input"
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          pattern={null}
          placeholder="Password"
          required={true}
          type="password"
          value=""
        />
      </div>
      <input
        className="submit"
        disabled={false}
        type="submit"
        value="Login"
      />
    </form>
  </div>
</div>
`;
